clvm/CLVMObject.py

```diff
@@ -1,47 +1,107 @@
-"""
-This is the minimal `SExp` type that defines how and where its contents are
-stored in the heap. The methods here are the only ones required for `run_program`.
-A native implementation of `run_program` should implement this base class.
-"""
+from typing import Any, Tuple, Optional, Union
+from blspy import G1Element
+from .casts import int_to_bytes
 
-import typing
 
-# Set this to "1" to do a run-time type check
-# This may slow things down a bit
+def looks_like_clvm_object(o: Any) -> bool:
+    d = dir(o)
+    return "atom" in d and "pair" in d
 
-TYPE_CHECK = 0
+
+CastableType = Union[
+    "CLVMObject",
+    bytes,
+    int,
+    None,
+    G1Element,
+    Tuple[Any, Any],
+]
 
 
 class CLVMObject:
+    """
+    This class implements the CLVM Object protocol, but supports deferring
+    conversion of python types until they are acceessed. For example, it can hold a
+    G1Element, which is converted to bytes the first time 'atom' is accessed. It can
+    hold a list, which is converted into a cons-list one element at a time, when
+    iterated.
+    """
+
+    # if this is not bytes, it will be lazily converted to bytes when accessed
+    atom_: Optional[Any]
 
-    atom: typing.Optional[bytes]
-    pair: typing.Optional[typing.Tuple["CLVMObject", "CLVMObject"]]
-    __slots__ = ["atom", "pair"]
+    # this is always a 2-tuple, but the elements may not have been converted to
+    # a CLVM object, they need to be when accessed. If an element is already a
+    # valid clvm object, it should be left alone. Otherwise it should be wrapped
+    # in CLVMObject to defer conversion when accessed
+    pair_: Optional[Tuple[Any, Any]]
+    __slots__ = ["atom_", "pair_"]
 
-    def __new__(class_, v: "SExpType"):
+    def __new__(class_, v: CastableType):
         if isinstance(v, CLVMObject):
             return v
-        if TYPE_CHECK:
-            type_ok = (
-                isinstance(v, tuple)
-                and len(v) == 2
-                and isinstance(v[0], CLVMObject)
-                and isinstance(v[1], CLVMObject)
-            ) or isinstance(v, bytes)
-            # uncomment next line for debugging help
-            # if not type_ok: breakpoint()
-            assert type_ok
+
+        # TODO: ideally we wouldn't mix SExp into the tree like this
+        if type(v).__name__ in ["SExp", "Program"]:
```

Comment:
is there a way to use `isinstance()` here, without importing `SExp`? As far as I understand it's not possible to import it here because you end up with a circular dependency

Reply:
Yeah, hard-coding `SExp` and `Program` here seems wrong. It might be possible to use `register` on an abstract base class as described here https://docs.python.org/3/library/abc.html but I haven't tested this.

Reply:
I think the source of this issue is that we create `CLVMObjects` from `SExp` sometimes, so we need to unwrap them here. if we would always use the higher level `SExp` to build trees, we would be fine since it can do the unwrapping. So I consider this temporary measure regardless